pkg/datapath/ipsec/ipsec_linux.go

```
@@ -0,0 +1,260 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// +build linux
+
+package ipsec
+
+import (
+	"net"
+
+	"github.com/cilium/cilium/pkg/datapath/linux"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/node"
+	"github.com/vishvananda/netlink"
+
+	"github.com/sirupsen/logrus"
+)
+
+type ipSecKeys struct {
+	Spi   int
+	ReqID int
+	Auth  *netlink.XfrmStateAlgo
+	Crypt *netlink.XfrmStateAlgo
+}
+
+var ipSecKeysGlobal = &ipSecKeys{
+	Spi:   1,
+	ReqID: 1,
+	Auth: &netlink.XfrmStateAlgo{
+		Name: "hmac(sha256)",
+		Key:  []byte("abcdefghijklmnopqrstuvwzyzABCDEF"),
+	},
+	Crypt: &netlink.XfrmStateAlgo{
+		Name: "cbc(aes)",
+		Key:  []byte("abcdefghijklmnopqrstuvwzyzABCDEF"),
+	},
+}
+
+func ipSecNewState() *netlink.XfrmState {
+	state := netlink.XfrmState{
+		Mode:  netlink.XFRM_MODE_TUNNEL,
+		Proto: netlink.XFRM_PROTO_ESP,
+		ESN:   false,
+	}
+	return &state
+}
+
+func ipSecNewPolicy() *netlink.XfrmPolicy {
+	policy := netlink.XfrmPolicy{}
+	return &policy
+}
+
+func ipSecAttachPolicyTempl(policy *netlink.XfrmPolicy, keys *ipSecKeys, srcIP net.IP, dstIP net.IP) {
+	tmpl := netlink.XfrmPolicyTmpl{
+		Proto: netlink.XFRM_PROTO_ESP,
+		Mode:  netlink.XFRM_MODE_TUNNEL,
+		Spi:   keys.Spi,
+		Reqid: keys.ReqID,
+		Dst:   dstIP,
+		Src:   srcIP,
+	}
+
+	policy.Tmpls = append(policy.Tmpls, tmpl)
+}
+
+func ipSecJoinState(state *netlink.XfrmState, keys *ipSecKeys) {
+	state.Auth = keys.Auth
+	state.Crypt = keys.Crypt
+	state.Spi = keys.Spi
+	state.Reqid = keys.ReqID
+}
+
+func ipSecGetLocalIP(endpointIP net.IP) net.IP {
+	if endpointIP.To4() != nil {
+		return node.GetInternalIPv4()
+	}
+	return node.GetIPv6Router()
+}
+
+func ipSecReplaceState(remoteIP, localIP net.IP) error {
+	state := ipSecNewState()
+
+	ipSecJoinState(state, ipSecKeysGlobal)
+	state.Src = localIP
+	state.Dst = remoteIP
+	err := netlink.XfrmStateAdd(state)
```

small nit you can return the error without assigning to a new variable

---

yeah I had some old debug code there to log the error but now just wasted line.